Mathlib/CategoryTheory/Monoidal/Comon_.lean

```diff
@@ -0,0 +1,173 @@
+/-
+Copyright (c) 2024 Lean FRO LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import Mathlib.CategoryTheory.Monoidal.Mon_
+import Mathlib.CategoryTheory.Monoidal.Opposite
+
+/-!
+# The category of comonoids in a monoidal category.
+
+We define comonoids in a monoidal category `C`.
+
+## TODO
+* `Comon_ C ≌ Mon_ (Cᵒᵖ)`
+* An oplax monoidal functor takes comonoid objects to comonoid objects.
+  That is, a oplax monoidal functor `F : C ⥤ D` induces a functor `Comon_ C ⥤ Comon_ D`.
+* Comonoid objects in `C` are "just"
+  oplax monoidal functors from the trivial monoidal category to `C`.
+* The category of comonoids in a braided monoidal category is monoidal.
+  (It may suffice to transfer this across the equivalent to monoids in the opposite category.)
+
+-/
+
+set_option linter.uppercaseLean3 false
+
+universe v₁ v₂ u₁ u₂ u
+
+open CategoryTheory MonoidalCategory
+
+variable (C : Type u₁) [Category.{v₁} C] [MonoidalCategory.{v₁} C]
+
+/-- A comonoid object internal to a monoidal category.
+
+When the monoidal category is preadditive, this is also sometimes called a "coalgebra object".
+-/
+structure Comon_ where
+  /-- The underlying object of a comonoid object. -/
+  X : C
+  /-- The counit of a comonoid object. -/
+  counit : X ⟶ 𝟙_ C
+  /-- The comultiplication morphism of a comonoid object. -/
+  comul : X ⟶ X ⊗ X
+  counit_comul : comul ≫ (counit ⊗ 𝟙 X) = (λ_ X).inv := by aesop_cat
+  comul_counit : comul ≫ (𝟙 X ⊗ counit) = (ρ_ X).inv := by aesop_cat
+  comul_assoc : comul ≫ (comul ⊗ 𝟙 X) = comul ≫ (𝟙 X ⊗ comul) ≫ (α_ X X X).inv := by aesop_cat
```

Comment:
By symmetry with the new definition of monoid objects, could you use the whiskering operators `▷` and `◁` instead of tensoring the identity?

Reply:
Done.

(And proof repair done by `rw_search`, nice!)